### PR TITLE
Implement lockorder checking on RwLocks in debug_sync

### PR DESCRIPTION
**Pull request by TheBlueMatt**

### CODE REVIEW

File: lightning/src/debug_sync.rs [hunk 1]
```diff
@@ -171,54 +180,109 @@ impl<T> Mutex<T> {
...
+	let _a = mutex.lock().unwrap();
+	let _b = mutex.lock().unwrap();
+}
+
+#[test]
+fn recursive_read() {
+	let mutex = RwLock::new(());
+	let _a = mutex.read().unwrap();
+	let _b = mutex.read().unwrap();
+}
```
Comment (jkczyz) [inline lightning/src/debug_sync.rs:273]: Could you add `should_panic` test cases for `write`/`write` and `write`/`read`?

File: lightning/src/debug_sync.rs [hunk 2]
```diff
@@ -64,6 +64,59 @@ impl std::hash::Hash for MutexMetadata {
...
 
+impl MutexMetadata {
+	fn new() -> MutexMetadata {
+		MutexMetadata {
+			locked_before: StdMutex::new(HashSet::new()),
+			mutex_idx: MUTEX_IDX.fetch_add(1, Ordering::Relaxed) as u64,
+			#[cfg(feature = "backtrace")]
+			mutex_construction_bt: Backtrace::new(),
+		}
+	}
```
Comment (jkczyz) [inline lightning/src/debug_sync.rs:75]: Perhaps uses of "mutex" should be renamed "lock"?

File: lightning/src/debug_sync.rs [hunk 3]
```diff
@@ -205,6 +213,7 @@ impl<T: Sized> Deref for RwLockReadGuard<'_, T> {
 
 impl<T: Sized> Drop for RwLockReadGuard<'_, T> {
 	fn drop(&mut self) {
+		if !self.first_lock { return; }
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/debug_sync.rs:216]: In theory, yea, in practice it should be exceedingly rare cause ~all our locks are RAII. I'll just add a comment noting that its a limitation.

**Review by arik-so — DISMISSED** [review]
seems fine as far as I can tell and test

**Review by jkczyz — APPROVED** [review]
Good to squash

**Comment by TheBlueMatt** [comment]
Squashed without change.

**Review by jkczyz — APPROVED** [review]

**Review by arik-so — APPROVED** [review]
Re-ACKing

**Pull request merged**